### PR TITLE
Mutation class #57

### PR DESCRIPTION
**Pull request by ericmjl**
This PR is ready for review. 

In this PR, we add in Mutation and MutationSet classes. The intent is to represent mutations made to a SeqLike. I've taken inspiration from multiple places, but the biggest one has been the discussion on #57. 

The use cases for this Mutation class are primarily two-fold:

1. Adding them to a SeqLike returns another mutated SeqLike.
2. Subtracting (diffing) a SeqLike from another yields a MutationSet.

While working through use case 2, I noticed how it's a bit tricky:

- adding a Mutation or MutationSet to a SeqLike results in a new SeqLike, 
- but subtracting the new SeqLike from the original SeqLike might not necessarily result in the original MutationSet, 
- yet adding the new MutationSet to the original SeqLike will give back a SeqLike identical to the new SeqLike.

An example of this phenomena is shown in a new notebook, `docs/notebooks/mutations.ipynb`. For reviewing purposes, that is probably the go-to notebook for understanding what Mutation and MutationSet classes can do; the rest are implementation details.

Would love to get feedback on this PR -- especially if there are semantics that I haven't yet thought of.

---

TODO list of what's left:

- [x] FIX TEST: SeqLike classes have two more class methods. Therefore, 77 is correct. Reference [here](https://github.com/modernatx/seqlike/actions/runs/3129446322/jobs/5078542482#step:5:92).
- [x] Add `.positions` class method, which returns a list of positions to mutate.
- [x] Switch out `magical_parse()` for `__new__()` under `Mutation`.

### CODE REVIEW

File: seqlike/SeqLikeAccessor.py
Comment (andrewgiessel) [inline seqlike/SeqLikeAccessor.py:65]: consider adding a `value_counts()` or `unique()`.  the output below could be really long

File: seqlike/MutationSet.py
Comment (andrewgiessel) [inline seqlike/MutationSet.py:13]: if you wrap this in enumerate, the ValueError below could print it.

File: seqlike/MutationSet.py
Comment (andrewgiessel) [inline seqlike/MutationSet.py:36]: out of date comment

File: seqlike/MutationSet.py
Comment (andrewgiessel) [inline seqlike/MutationSet.py:57]: What does it mean to add an integer to a MutationSet?  The constructor only takes a iterator of Mutations or strs, not ints, so this isn't clear as is

File: seqlike/SeqLike.py
Comment (andrewgiessel) [inline seqlike/SeqLike.py:37]: this can be imported from Mutation directly

File: seqlike/Mutation.py
Comment (andrewgiessel) [inline seqlike/Mutation.py:103]: I think this should be a method of Mutation- as far as I can tell, it's only called in its constructor.  In the spirit of "There should be one– and preferably only one –obvious way to do it", I think it makes sense to make the obvious thing instantiating a Mutation object.

File: seqlike/SeqLike.py
Comment (ericmjl) [reply seqlike/SeqLike.py:37]: That doesn't "feel" right, I think -- doing so feels like an indirect import. In this case, an additional line doesn't feel like it would be of too much consequence, and would also have the advantage of being explicit about where the MutationSet class is defined.

File: seqlike/Mutation.py
Comment (ericmjl) [reply seqlike/Mutation.py:110]: I thought about this comment, and tried to retrace my thought process in keeping this function separate. The original idea, I think, was to keep the `Mutation.__init__()` clean, hence the contents of the function were kept separate from `__init__()`. I also desired to be able to call on the function in a standalone fashion if possible. Not sure why, but it felt like a sane decision to take. I guess we can make it a class method, but I think that would make calling on `parse` standalone a bit more difficult?

File: seqlike/MutationSet.py
Comment (ericmjl) [reply seqlike/MutationSet.py:71]: I've addressed it in the docstring. Essentially "offset the position".

File: seqlike/SeqLike.py
Comment (andrewgiessel) [reply seqlike/SeqLike.py]: Sorry, this comment was out of date after you got rid of magical_parse.

File: seqlike/Mutation.py
Comment (andrewgiessel) [reply seqlike/Mutation.py:110]: Stand-alone or not is a matter of taste I suppose.  I'd rather make Mutations and use attributes to get what I need, but I since this pattern is currently supported, I'm ok leaving it as a standalone function.  I agree that it could be useful.

However, I'd **strongly** recommend changing the name to something more specific than "parse".  How many parse functions are there in the world? =)

**Comment by ericmjl** [comment]
I am going to do a few more commits to address @andrewgiessel's remaining comments.

**Comment by ndousis** [comment]
@ericmjl , this new mutation functionality is awesome, and the jupyter notebook is very helpful in exploring it!  The notation and "arithmetic" conventions are simple, clear and sensible.

The following behavior doesn't look right, however:

```python
s1 = aaSeqLike("MKAILV")
ms2 = MutationSet(mutations="^2A;^4D;5-".split(";"))
str(s1 + ms2)
```
yields
```
MAAD-V
```
whereas I'd expect
```
MAKD-ILV
```

Perhaps this is related, but the following also gives an unexpected result:
```python
s1 - (s1 + Mutation("5-"))
```
yields
```
[5V, 6-]
```
instead of
```
[5-]
```

**Comment by ericmjl** [comment]
@ndousis thanks for the feedback! 

The behaviour you've outlined is one that I've been thinking about w.r.t. semantics. The essence is that there is an incrementing of position numbers every time that we have an insertion, so that any insertions are preserved relative to the _original_ numbering of the sequence. In the example you showed, on the latest commit `ffdb0bc`, I get the behaviour defined by the aforementioned semantics:

image

Do you think it'll help to jump on a call with you to see if the current semantics make sense? If so, could you email me at my work address (intentionally not posting here to avoid spam)? I'll send you a calendly link that we can use to schedule.

**Comment by ericmjl** [comment]
@ndousis and @andrewgiessel this PR is ready for more reviews. I also hope that we can get official "approvals" on the PR before merging, so that we leave a record.

File: seqlike/Mutation.py
Comment (ericmjl) [inline seqlike/Mutation.py:99]: Add comment on why Substitution and Deletion are empty classes.

File: seqlike/Mutation.py
Comment (ericmjl) [inline seqlike/Mutation.py:117]: Move this to the top.

**Review by andrewgiessel — APPROVED** [review]
Approving, pending final discussion/clarification on MutationSet Semantics + resulting documentation

**Comment by ndousis** [comment]
@ericmjl , I think the current semantics make sense, and they do yield stable solutions.  I was troubled by the asymmetry of cases like `s1 - (s1 + Mutation("5-"))`, but as you've pointed out in a previous version of `mutations.ipynb`, consistent annotation of indels is difficult without pairwise alignment, which in itself is inconsistent.

Mutation and MutationSet are very thoughtful and useful new features in SeqLike. Approved!

**Review by ndousis — APPROVED** [review]

**Comment by ericmjl** [comment]
Thank you, @andrewgiessel and @ndousis! I am going to merge and cut a new release once the CI passes!

**Pull request merged**
